### PR TITLE
fix: Add Rate limit to Broadcast endpoint

### PR DESCRIPTION
**Pull request by filipecabaco**
## What kind of change does this PR introduce?

To prevent over usage by customers, we will take into account the Rate Limits.

We handle two situations:
* The existing batch will trigger a rate limit so we advice to reduce the batch size
* Rate limit has been hit so we can't process the request

### CODE REVIEW

**Review by abc3 — APPROVED** [review]
LGTM, just want to say that there are failed tests

**Comment by chasers** [comment]
Mmmm ... we're already have this logic in the rate limiter plug `RealtimeWeb.Plugs.RateLimiter`.

We should be doing this there.

File: lib/realtime_web/controllers/broadcast_controller.ex [hunk 1]
```diff
@@ -30,26 +33,45 @@ defmodule RealtimeWeb.BroadcastController do
...
+         :ok <- check_rate_limit(events_per_second_key, tenant, length(messages)) do
       for %{changes: %{topic: sub_topic, payload: payload, event: event}} <- messages do
         tenant_topic = Tenants.tenant_topic(tenant, sub_topic)
         payload = %{"payload" => Map.merge(payload, %{"event" => event})}
 
         Endpoint.broadcast_from(self(), tenant_topic, "broadcast", payload)
 
         GenCounter.add(events_per_second_key)
-        Logger.info("Broadcasted message to using HTTP API: " <> tenant_topic)
+        Logger.info("Broadcasted message using HTTP API to topic " <> tenant_topic)
```
Comment (chasers) [inline lib/realtime_web/controllers/broadcast_controller.ex:54]: Thought you were only adding this for testing. We need to pull it out we don't want every Broadcast request being logged here.

File: lib/realtime_web/controllers/broadcast_controller.ex [hunk 2]
```diff
@@ -30,26 +33,45 @@ defmodule RealtimeWeb.BroadcastController do
...
       end
 
       send_resp(conn, :accepted, "")
     end
   end
+
+  defp check_rate_limit(events_per_second_key, %Tenant{} = tenant, total_messages_to_broadcast) do
```
Comment (chasers) [inline lib/realtime_web/controllers/broadcast_controller.ex:61]: Can we consolidate this logic in `RealtimeWeb.Plugs.RateLimiter`?

**Comment by filipecabaco** [comment]
TIL we had `RealtimeWeb.Plugs.RateLimiter` 

That will make it trickier to check the batch one 🤔 will refactor now

**Comment by chasers** [comment]
If it comes after the JSON parser we should have the JSON list in the conn to count like you were doing.

**Comment by filipecabaco** [comment]
Sent a message to Slack regarding this discussion since it might require more checks.

Issue is that we might want to use a different rate limit key

**Comment by filipecabaco** [comment]
For public record sake 😄 

The keys we're using to check rate limiting on broadcast isn't the same as the rate limiter plug as such we can't properly merge the logic as expected.

We will revise the Plug logic and this endpoints logic to better abstract the rate limit keys at a later stage

**Pull request merged**